cumulusci/robotframework/utils.py

```diff
@@ -199,3 +200,18 @@ def run_with_retry(self, *args, **kwargs):
     else:
         # Decorator was used without arguments
         return decorate(target)
+
+
+def capture_screenshot_on_error(func):
+    """Decorator for capturing a screenshot if a keyword throws an error"""
```

Comment:
It seems to me that this would be more helpful if we can turn on capturing a screenshot after an error as globally as possible, rather than just decorating specific keywords. With the decorator, it seems inevitable that people will run into situations where the error happened in a keyword that didn't have the decorator, and then extra effort is required to add the decorator and reproduce the issue.

My suggestion would be that we recommend that test authors do this at the top of any test suite that uses selenium:

    Test Teardown | Run Keyword If Test Failed | Capture Page Screenshot

(Or maybe there's even a way to make that be a default test teardown if it's not set?)

Reply:
I don't think we want to raise it to a more global level. Not all keywords should capture screenshots. API keywords, for example, don't need to have screenshots on failure. 

Also, the test teardown is problematic, as that code may not run immediately after the error occurs.  Time may pass, or other code may run before the screenshot is taken. It's a good safety net, but it's not as effective as capturing the screenshot immediately after an error has occurred.  Plus, by using the decorator the screenshot shows up in the log immediately before the error message without the user having to drill down into a teardown to see it. 

That being said, automatically taking a screenshot without the user having to do anything at all has merit. We could easily add that to a listener that we automatically enable when we run tests.

Reply:
That's a fair point that screenshots are not relevant to all keywords. (Though if a browser was open when an error happened in a non-selenium keyword, I don't think it would hurt to take the screenshot anyway to provide extra context.)

I don't understand the point about a delay before test teardown. Won't an error cause the test to end so that test teardown is the very next thing that happens?

I haven't thought through the details but I like the idea of using a listener. It seems like that might give us the ability to be a bit more selective but also avoid needing to decorate lots of separate keywords.

By the way, there is a previous attempt I made to automatically take screenshots whenever there's an exception in a selenium command. That is here: https://github.com/SFDO-Tooling/CumulusCI/blob/master/cumulusci/robotframework/utils.py#L88 ...this worked, but I sometimes wished that I would get a screenshot even for exceptions that didn't occur during selenium commands.

Reply:
> I don't understand the point about a delay before test teardown. Won't an error cause the test to end so that test teardown is the very next thing that happens?

Not necessarily. In the normal case it would, but tests could call [Run keyword and continue on failure](http://robotframework.org/robotframework/latest/libraries/BuiltIn.html#Run%20Keyword%20And%20Continue%20On%20Failure) which lets you call additional keywords even after the test has technically failed. We do that in one of our core tests so that we can make multiple assertions rather than stopping at the first failure.

Reply:
I'll think a bit more about the listener idea. My gut tells me there's some edge case that I'm forgetting about, but maybe I'm just hungry :-)

Reply:
Okay that makes sense. I didn't know about `Run keyword and continue on failure`.

Reply:
Did you think more about the listener idea? I'd rather not add the decorator if we're just going to replace it.

Reply:
I think I prefer keeping the decorator. I think it gives more control to our keyword authors.

Reply:
Ugh. I didn't realize we did that. I'll look into it.

Reply:
I want to keep the decorator. After doing some investigation I learned that there are many cases where our automatic screenshot code won't capture screenshots, since it only catches exceptions thrown by the selenium module rather than SeleniumLibrary.

Because it doesn't catch everything, I need an easy way to capture screenshots on failure. I can get rid of the decorator and make liberal use of try/except, but I found myself duplicating that pattern over and over which is why I created the decorator in the first place. 

With respect to the concern over duplicate screenshots, at the moment that chance is zero since the page objects aren't using the retrying selenium decorator.  Maybe they should, but they are working fine without it, and do things that that code wouldn't catch.

Reply:
@boakley I'm thinking that in the end it would be best if we did all of the following:
1. Add this decorator, as a way to do targeted screenshots when particular keywords fail.
2. Add a listener that takes a screenshot after any keyword fails _if_ a browser is open. This gives an option for users who want easy visibility into failures without needing to track down a specific place to add the decorator and then re-run the tests.
3. Remove the current screenshot capture that's monkeypatched in by the selenium_retry decorator. It's too tightly coupled to the retry decorator, untested, and doesn't catch all the situations where a screenshot is desired.

In the interest of getting this PR merged, I'd suggest we do the first one now and follow up later on 2 and 3. Does that sound okay to you?

Reply:
I agree on points 1 and 3, point 2 needs a bit more thought. There are times when we won't want a keyword to generate a screenshot on error, but perhaps we can come up with a solution that addresses that.